File: producers/python/airep_producer.py
#!/usr/bin/env python3
"""AIREP record producer (Python) — sign your OWN record with your OWN key.

This is the copy-paste path for emitting a conformant AIREP v0.1 record from your runtime.
It fills the ``integrity`` block exactly as the conformance verifiers re-derive it:

    body      = the record with integrity.current and integrity.signature removed
                (integrity.previous and canonical_json are RETAINED)
    current   = "sha256:" + SHA-256( jcs.canonicalize(body) )    # RFC 8785 canonical bytes
    signature = Ed25519 signature over the `current` string, hex-encoded

It signs with a CALLER-SUPPLIED key, never the published test key. To use it in your own
project, copy two files into it — this file and ``spec/airep/v0.1/conformance/jcs.py``
(stdlib-only) — and change the ``import jcs`` line below to point at your copy.

Quickstart (see EXPLAINER.md "Write your own record in 5 minutes"):

    # draft a record (copy an example and edit the fields), then sign it with a fresh key:
    python3 producers/python/airep_producer.py my_record.json
    # -> fills integrity in place, writes my_record.pub.hex + my_record.key.hex, and prints
    #    the exact verify command (you will see `sig=ok`).

Requires: cryptography (pip install cryptography).
"""
from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

# The AIREP RFC 8785 canonicalizer. In-repo it lives in the conformance kit; a producer in
# another project copies jcs.py next to this file and uses `import jcs` directly.
_CONF = Path(__file__).resolve().parents[2] / "spec" / "airep" / "v0.1" / "conformance"
if str(_CONF) not in sys.path:
    sys.path.insert(0, str(_CONF))
import jcs  # noqa: E402  (RFC 8785 canonicalization; see conformance/jcs.py)

GENESIS = "sha256:" + "0" * 64


def build_record(record: dict, signing_key: Ed25519PrivateKey, previous: str = GENESIS) -> dict:
    """Return ``record`` with a complete ``integrity`` block.

    The hash is computed over the record's RFC 8785 canonical form with ``integrity.current``
    and ``integrity.signature`` removed and ``integrity.previous`` retained, so content is bound
    to chain position. ``previous`` defaults to the genesis constant; pass the prior record's
    ``integrity.current`` to extend a chain.
    """
    out = dict(record)
    integ = dict(out.get("integrity", {}))
    integ["previous"] = previous
    integ["canonical_json"] = True

    body = dict(out)
    body["integrity"] = {k: v for k, v in integ.items() if k not in ("current", "signature")}
    current = "sha256:" + hashlib.sha256(jcs.canonicalize(body)).hexdigest()

    integ["current"] = current
    integ["signature"] = {"alg": "Ed25519", "value": signing_key.sign(current.encode("utf-8")).hex()}
    out["integrity"] = integ
    return out


def _pub_hex(sk: Ed25519PrivateKey) -> str:
    return sk.public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw).hex()


def main() -> int:
    ap = argparse.ArgumentParser(
        description="Fill an AIREP record's integrity block and sign it with your own Ed25519 key.")
    ap.add_argument("record", help="JSON record you drafted (subject/input/claim/output/evidence/"
                                    "directive/scope); copy an example from examples/ and edit it")
    ap.add_argument("--key", metavar="HEXFILE",
                    help="file holding a 32-byte Ed25519 seed in hex; a fresh key is generated if omitted")
    ap.add_argument("--previous", default=GENESIS,
                    help="integrity.previous — the prior record's current, or genesis (default)")
    args = ap.parse_args()

    generated = args.key is None
    if generated:
        sk = Ed25519PrivateKey.generate()
    else:
        sk = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(Path(args.key).read_text().strip()))

    rec_path = Path(args.record)
    signed = build_record(json.loads(rec_path.read_text()), sk, previous=args.previous)
    rec_path.write_text(json.dumps(signed, indent=2) + "\n")

    pub = _pub_hex(sk)
    pub_path = rec_path.with_suffix(".pub.hex")
    pub_path.write_text(pub + "\n")

    print(f"signed            -> {rec_path}")
    print(f"integrity.current  = {signed['integrity']['current']}")
    print(f"public key (hex)   -> {pub_path}")
    if generated:
        seed = sk.private_bytes(serialization.Encoding.Raw, serialization.PrivateFormat.Raw,
                                serialization.NoEncryption()).hex()
        key_path = rec_path.with_suffix(".key.hex")
        key_path.write_text(seed + "\n")
        key_path.chmod(0o600)
        print(f"NEW private key    -> {key_path}  (chmod 600 — keep it out of git and out of records)")
    print("\nverify it:")
    print(f"  python3 {_CONF}/verify.py {rec_path} --pubkey {pub_path} --class")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: producers/typescript/airep_producer.ts
#!/usr/bin/env node
// SPDX-License-Identifier: Apache-2.0
/**
 * AIREP v0.1 — an independent PRODUCER in TypeScript (zero dependencies; node:crypto only).
 *
 * Why this exists: AIREP ships two independent *verifiers* (Python + Node) that agree byte-for-byte.
 * Verifiers are readers. This is a *writer* on a third, independent stack — it emits a signed,
 * hash-chained AIREP chain that BOTH reference verifiers accept byte-for-byte. A format whose records
 * can be produced outside its author's Python runtime, and still validate, is an interchange format
 * rather than one tool's serialization. (Author-written: a second language/stack, same author — a
 * bridge toward, not a substitute for, a truly third-party producer; see ../../spec/airep/v0.1/STATUS.md.)
 *
 *   tsc airep_producer.ts && node airep_producer.js > /tmp/ts_chain.jsonl
 *   python3 ../../spec/airep/v0.1/conformance/verify.py  /tmp/ts_chain.jsonl --pubkey <pubkey>
 *   node     ../../spec/airep/v0.1/conformance/verify.mjs /tmp/ts_chain.jsonl --pubkey <pubkey>
 */
import { createHash, createPrivateKey, sign as edSign, KeyObject } from "node:crypto";

const GENESIS = "sha256:" + "0".repeat(64);

// RFC 8785 (JCS) canonical form. This is byte-identical to the reference verify.mjs canonicalizer,
// which agrees byte-for-byte with the Python jcs.py over a value battery (conformance/test_jcs.py):
// recursive key sort + native ES6 JSON.stringify (the number form JCS mandates) + no whitespace.
// Records here carry only strings and integers (no floats), so JSON.stringify == JCS exactly.
function canonical(v: unknown): string {
  if (v === null) return "null";
  if (Array.isArray(v)) return "[" + v.map(canonical).join(",") + "]";
  if (typeof v === "object") {
    const o = v as Record<string, unknown>;
    return "{" + Object.keys(o).sort().map((k) => JSON.stringify(k) + ":" + canonical(o[k])).join(",") + "}";
  }
  return JSON.stringify(v); // string | number | boolean
}

function sha256(text: string): string {
  return "sha256:" + createHash("sha256").update(text, "utf8").digest("hex");
}

// The published THROWAWAY test seed (TEST ONLY — never use for any real signing identity). It derives
// the published examples/test_public_key.txt, so a chain signed here verifies under that public key.
const TEST_SEED_HEX = "00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff";
function testPrivateKey(): KeyObject {
  // PKCS8 DER for an Ed25519 private key = fixed 16-byte prefix + the 32-byte seed.
  const pkcs8 = Buffer.concat([
    Buffer.from("302e020100300506032b657004220420", "hex"),
    Buffer.from(TEST_SEED_HEX, "hex"),
  ]);
  return createPrivateKey({ key: pkcs8, format: "der", type: "pkcs8" });
}

interface Evidence { type: string; ref: string; resolvable: boolean; content_hash?: string }
interface Decision {
  verb: "release" | "block" | "defer" | "redact" | "escalate_to_human" | "kill";
  assertion: string;
  basis: string[];
  evidence: Evidence[];
  covers: string[];
  doesNotCover: string[];
}

interface AirepRecord {
  airep_version: "0.1";
  subject: { runtime: string; producer: string; decision_index: number; timestamp_utc: string };
  input: { input_ref: string; governance_state: Record<string, unknown> };
  claim: { assertion: string; basis: string[] };
  output: { result_ref: string; redacted?: boolean };
  evidence: Evidence[];
  directive: { verb: string; policy_basis: string[] };
  scope: { covers: string[]; does_not_cover: string[] };
  integrity: { previous: string; canonical_json: true; current?: string; signature?: { alg: string; value: string } };
}

function buildRecord(d: Decision, index: number, previous: string, priv: KeyObject): AirepRecord {
  const rec: AirepRecord = {
    airep_version: "0.1",
    subject: {
      runtime: "ts-reference-producer",
      producer: "airep-producer-ts/0.1",
      decision_index: index,
      timestamp_utc: "2026-05-31T00:00:00Z",
    },
    input: { input_ref: sha256(`input-${index}`), governance_state: { policy_version: "p1" } },
    claim: { assertion: d.assertion, basis: d.basis },
    output: { result_ref: sha256(`output-${index}`), redacted: d.verb === "redact" },
    evidence: d.evidence,
    directive: { verb: d.verb, policy_basis: d.basis },
    scope: { covers: d.covers, does_not_cover: d.doesNotCover },
    // hash is computed with integrity = {previous, canonical_json} (current + signature absent),
    // exactly the body the verifiers recompute (current + signature removed, previous retained).
    integrity: { previous, canonical_json: true },
  };
  const current = sha256(canonical(rec));
  rec.integrity.current = current;
  const value = edSign(null, Buffer.from(current, "utf8"), priv).toString("hex");
  rec.integrity.signature = { alg: "Ed25519", value };
  return rec;
}

const DECISIONS: Decision[] = [
  {
    verb: "release", assertion: "answer released after the safety and ethics gates passed",
    basis: ["safety_gate", "ethics_gate"],
    evidence: [{ type: "policy", ref: "policy://safety/v1", content_hash: sha256("safety/v1"), resolvable: true }],
    covers: ["the safety and ethics gates fired and passed"],
    doesNotCover: ["whether the answer is factually correct", "reasoning faithfulness not attested"],
  },
  {
    verb: "block", assertion: "output blocked by the ethics gate",
    basis: ["ethics_gate"],
    evidence: [{ type: "policy", ref: "policy://ethics/v2", content_hash: sha256("ethics/v2"), resolvable: true }],
    covers: ["the ethics gate fired and blocked the output"],
    doesNotCover: ["whether a human would have blocked it", "factual correctness not assessed"],
  },
  {
    verb: "escalate_to_human", assertion: "handed to a human reviewer after a low-confidence signal",
    basis: ["hitl_queue"],
    evidence: [{ type: "human_approval", ref: "hitl://ticket/42", resolvable: false, content_hash: sha256("ticket/42") }],
    covers: ["the decision was queued for human review"],
    doesNotCover: ["the human's eventual verdict", "factual correctness not assessed"],
  },
];

function main(): void {
  const priv = testPrivateKey();
  let previous = GENESIS;
  const lines: string[] = [];
  DECISIONS.forEach((d, i) => {
    const rec = buildRecord(d, i, previous, priv);
    lines.push(JSON.stringify(rec));
    previous = rec.integrity.current as string;
  });
  process.stdout.write(lines.join("\n") + "\n");
}

main();
